### DISCUSSION TITLE
Base strategy adaptation for revisiting the exact same data with self.batches

### DISCUSSION THREAD

**Discussion by Mattdl** [ContinualAI/avalanche#451] [Feature Request]
Hi! I'm trying to finetune an independent model within a strategy on the exact same data as seen for the current batch.
However, the losses for this model I want to add to the loss of the model being optimized currently in the base strategy.

In practice, this proves difficult as before/after_forward happen before the calculating the loss fo the current network. But after `loss+=criterion(...)` for the current task I would like to process the loss for the independently finetuned model, so I can just add to the loss as well within the strategy.

What I tried now is to implement it in the before_backward, and to iterate over the exact same data:
`    def before_backward(self, strategy, **kwargs):
             for mb_task_id, (mb_x, mb_y) in strategy.current_dataloader[strategy.mb_it].items():
                  .... # Process
`
However, I get an error because I can't index the dataloader. I think this would be solved if there was a self.batches in the base strategy which allows to easily iterate over the batches again.
This would allow to:
`    def before_backward(self, strategy, **kwargs):
             for mb_task_id, (mb_x, mb_y) in self.batches.items():
                  .... # Process
`

**Comment by AntonioCarta** [comment]
If I understand correctly your problem, you may be able to solve it by modifying also the `criterion`? 
Either way, I think you are right that we need to provide access to `batches`. That part of the `BaseStrategy` was written before `AvalancheDataset` supported multi-task data.

I will work on this (#452).

**Reply by Mattdl** [reply]
Thanks for the quick reply! So to be clear, the `batches` (and `self.mb_it`) refer to minibatches for multiple tasks for a single round of processing? For me this was not directly clear from the code, so might be interesting to put in the `Base_Strategy` documentation (: 
Anyway, in that case, I only have 1 batch in `batches`, so I can just use the `self.mb_x` to solve my problem in the `before_backward`.

Thanks for the help!

**Reply by AntonioCarta** [reply]
yes, you are correct. It is only needed to average the loss among multiple tasks before doing the backpropagation. I will add more documentation about the internal state of the `BaseStrategy` soon.
